cloud/pkg/devicecontroller/controller/downstream.go

```diff
@@ -438,6 +438,103 @@ func isNodeSelectorUpdated(old *v1.NodeSelector, new *v1.NodeSelector) bool {
+	return !reflect.DeepEqual(old, new)
+}
+
+// updateProtocolInConfigMap updates the protocol in the deviceProfile in configmap
+func (dc *DownstreamController) updateProtocolInConfigMap(device *v1alpha1.Device) {
+	if len(device.Spec.NodeSelector.NodeSelectorTerms) != 0 && len(device.Spec.NodeSelector.NodeSelectorTerms[0].MatchExpressions) != 0 && len(device.Spec.NodeSelector.NodeSelectorTerms[0].MatchExpressions[0].Values) != 0 {
+		configMap, ok := dc.configMapManager.ConfigMap.Load(device.Spec.NodeSelector.NodeSelectorTerms[0].MatchExpressions[0].Values[0])
+		if ok {
+			nodeConfigMap, ok := configMap.(*v1.ConfigMap)
+			if !ok {
+				log.LOGGER.Errorf("Failed to assert to configmap")
+				return
+			}
+			dp, ok := nodeConfigMap.Data[DeviceProfileJSON]
+			if !ok || dp == "{}" {
+				// This case should never be hit as we delete empty configmaps
+				log.LOGGER.Errorf("Failed to get deviceProfile from configmap data or deviceProfile is empty")
+				return
+			}
+
+			deviceProfile := &types.DeviceProfile{}
+			err := json.Unmarshal([]byte(dp), deviceProfile)
+			if err != nil {
+				log.LOGGER.Errorf("Failed to unmarshal due to error: %v", err)
+				return
+			}
+			var oldProtocol string
+			for _, devInst := range deviceProfile.DeviceInstances {
+				if device.Name == devInst.Name {
+					oldProtocol = devInst.Protocol
+					break
+				}
+			}
+
+			// delete the old protocol
+			for i, ptcl := range deviceProfile.Protocols {
+				if ptcl.Name == oldProtocol {
+					deviceProfile.Protocols[i] = deviceProfile.Protocols[len(deviceProfile.Protocols)-1]
+					deviceProfile.Protocols[len(deviceProfile.Protocols)-1] = nil
+					deviceProfile.Protocols = deviceProfile.Protocols[:len(deviceProfile.Protocols)-1]
+				}
+			}
+
+			// add new protocol
+			deviceProtocol := &types.Protocol{}
+			var protocol string
+			if device.Spec.Protocol.OpcUA != nil {
+				protocol = OPCUA + "-" + device.Name
+				deviceProtocol.Name = protocol
+				deviceProtocol.Protocol = OPCUA
+				deviceProtocol.ProtocolConfig = device.Spec.Protocol.OpcUA
+			} else if device.Spec.Protocol.Modbus != nil && device.Spec.Protocol.Modbus.RTU != nil {
+				protocol = ModbusRTU + "-" + device.Name
+				deviceProtocol.Name = protocol
+				deviceProtocol.Protocol = ModbusRTU
+				deviceProtocol.ProtocolConfig = device.Spec.Protocol.Modbus.RTU
+			} else if device.Spec.Protocol.Modbus != nil && device.Spec.Protocol.Modbus.TCP != nil {
+				protocol = ModbusTCP + "-" + device.Name
+				deviceProtocol.Name = protocol
+				deviceProtocol.Protocol = ModbusTCP
+				deviceProtocol.ProtocolConfig = device.Spec.Protocol.Modbus.TCP
+			} else if device.Spec.Protocol.Bluetooth != nil {
+				protocol = Bluetooth + "-" + device.Name
+				deviceProtocol.Name = protocol
+				deviceProtocol.Protocol = Bluetooth
+				deviceProtocol.ProtocolConfig = device.Spec.Protocol.Bluetooth
+			} else {
+				log.LOGGER.Warnf("Unsupported device protocol")
```

Comment:
Is there a check before this routine to add only supported devices? If yes, then above default case is fine. If not, then it should return here with possible rollover of any previous configuration done.

Reply:
There is no check before this routine. the protocol can be empty but it should not stop updation of configmap. It will just delete the old protocol config as it is not associated with the device anymore.